rllib/agents/ppo/ddppo.py

```diff
@@ -89,186 +91,186 @@
+                            batch.count, config["sgd_minibatch_size"]))
+            info = do_minibatch_sgd(batch, this_worker.policy_map, this_worker,
+                                    config["num_sgd_iter"],
+                                    config["sgd_minibatch_size"],
+                                    ["advantages"])
+            return info, batch.count
+
+        # Broadcast the local set of global vars.
+        def update_worker_global_vars(item):
+            global_vars = _get_global_vars()
+            for w in workers.remote_workers():
+                w.set_global_vars.remote(global_vars)
             return item
 
-        train_op = train_op.for_each(download_weights)
-
-    # In debug mode, check the allreduce successfully synced the weights.
-    if logger.isEnabledFor(logging.DEBUG):
-
-        def check_sync(item):
-            weights = ray.get(
-                [w.get_weights.remote() for w in workers.remote_workers()])
-            sums = []
-            for w in weights:
-                acc = 0
-                for p in w.values():
-                    for k, v in p.items():
-                        acc += v.sum()
-                sums.append(float(acc))
-            logger.debug("The worker weight sums are {}".format(sums))
-            assert len(set(sums)) == 1, sums
-
-        train_op = train_op.for_each(check_sync)
-
-    return StandardMetricsReporting(train_op, workers, config)
-
-
-# Build a child class of `Trainer`, based on PPOTrainer's setup.
-# Note: The generated class is NOT a sub-class of PPOTrainer, but directly of
-# the `Trainer` class.
-DDPPOTrainer = ppo.PPOTrainer.with_updates(
-    name="DDPPO",
-    default_config=DEFAULT_CONFIG,
-    validate_config=validate_config,
-    execution_plan=execution_plan,
-)
+        # Have to manually record stats since we are using "raw" rollouts mode.
+        class RecordStats:
+            def _on_fetch_start(self):
+                self.fetch_start_time = time.perf_counter()
+
+            def __call__(self, items):
+                for item in items:
+                    info, count = item
+                    metrics = _get_shared_metrics()
+                    metrics.counters[STEPS_TRAINED_THIS_ITER_COUNTER] = count
+                    metrics.counters[STEPS_SAMPLED_COUNTER] += count
+                    metrics.counters[STEPS_TRAINED_COUNTER] += count
+                    metrics.info[LEARNER_INFO] = info
+                # Since SGD happens remotely, the time delay between fetch and
+                # completion is approximately the SGD step time.
+                metrics.timers[LEARN_ON_BATCH_TIMER].push(
+                    time.perf_counter() - self.fetch_start_time)
+
+        train_op = (
+            rollouts.for_each(train_torch_distributed_allreduce)  # allreduce
+            .batch_across_shards()  # List[(grad_info, count)]
+            .for_each(RecordStats()))
+
+        train_op = train_op.for_each(update_worker_global_vars)
```

Comment:
minor minor question, maybe just chain this call right above?

Reply:
Didn't actually touch this code, just moved it into the method. Not sure about too much chaining. Honestly, we should probably chain rather less than more as it makes the already complex execution plans even more confusing.

Reply:
ok, I can't argue against it 😆